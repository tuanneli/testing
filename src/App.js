import './App.css';
import {useEffect, useState} from "react";
import Users from "./components/users/Users";
import {BrowserRouter, Link, Route, Routes} from "react-router-dom";
import HomePage from "./pages/HomePage";
import AboutPage from "./pages/AboutPage";
import ErrorPage from "./pages/ErrorPage";
import UserDetailsPage from "./pages/UserDetailsPage";
import AppRouter from "./components/AppRouter";
import Navbar from "./components/navbar/Navbar";

function App() {
    const [data, setData] = useState(null);
    const [toggle, setToggle] = useState(false);
    const [value, setValue] = useState("");

    const onClick = () => setToggle(prev => !prev);

    useEffect(() => {
        setTimeout(() => {
            setData({});
        }, 100);
    }, []);

    return (
        <div className="App">
            <Navbar/>
            <AppRouter/>
        </div>
    );
}

export default App;
//
// class MyPromise {
//     constructor(executor) {
//         this.status = 'pending';
//         this.value = undefined;
//         this.reason = undefined;
//         this.onResolvedCallbacks = [];
//         this.onRejectedCallbacks = [];
//
//         const resolve = value => {
//             if (this.status === 'pending') {
//                 this.status = 'fulfilled';
//                 this.value = value;
//                 this.onResolvedCallbacks.forEach(cb => cb());
//             }
//         };
//
//         const reject = reason => {
//             if (this.status === 'pending') {
//                 this.status = 'rejected';
//                 this.reason = reason;
//                 this.onRejectedCallbacks.forEach(cb => cb());
//             }
//         };
//
//         try {
//             executor(resolve, reject);
//         } catch (error) {
//             reject(error);
//         }
//     }
//
//     then(onFulfilled, onRejected) {
//         onFulfilled = typeof onFulfilled === 'function' ? onFulfilled : value => value;
//         onRejected = typeof onRejected === 'function' ? onRejected : reason => {
//             throw reason
//         };
//
//         const promise2 = new Promise((resolve, reject) => {
//             if (this.status === 'fulfilled') {
//                 setTimeout(() => {
//                     try {
//                         const x = onFulfilled(this.value);
//                         resolvePromise(promise2, x, resolve, reject);
//                     } catch (error) {
//                         reject(error);
//                     }
//                 });
//             }
//
//             if (this.status === 'rejected') {
//                 setTimeout(() => {
//                     try {
//                         const x = onRejected(this.reason);
//                         resolvePromise(promise2, x, resolve, reject);
//                     } catch (error) {
//                         reject(error);
//                     }
//                 });
//             }
//
//             if (this.status === 'pending') {
//                 this.onResolvedCallbacks.push(() => {
//                     setTimeout(() => {
//                         try {
//                             const x = onFulfilled(this.value);
//                             resolvePromise(promise2, x, resolve, reject);
//                         } catch (error) {
//                             reject(error);
//                         }
//                     });
//                 });
//                 this.onRejectedCallbacks.push(() => {
//                     setTimeout(() => {
//                         try {
//                             const x = onRejected(this.reason);
//                             resolvePromise(promise2, x, resolve, reject);
//                         } catch (error) {
//                             reject(error);
//                         }
//                     });
//                 });
//             }
//         });
//
//         return promise2;
//     }
//
//     catch(onRejected) {
//         return this.then(null, onRejected);
//     }
// }
//
// const resolvePromise = (promise2, x, resolve, reject) => {
//     if (x === promise2) {
//         return reject(new TypeError('Circular reference'));
//     }
//
//     let called = false;
//
//     if (x instanceof Promise) {
//         x.then(value => {
//             resolvePromise(promise, value, resolve, reject);
//         }, reason => {
//             if (!called) {
//                 called = true;
//                 reject(reason);
//             }
//         });
//     } else {
//         resolve(x);
//     }
// };
//
// module.exports = Promise;
