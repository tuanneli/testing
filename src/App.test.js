import {fireEvent, render, screen} from '@testing-library/react';
import App from './App';
import userEvent from "@testing-library/user-event";

describe('Test App', () => {
    test('renders learn react link', () => {
        render(<App/>);
        const helloElement = screen.getByText(/Hello there/i);
        const btn = screen.getByRole('button');
        const input = screen.getByPlaceholderText(/input value/i)
        expect(helloElement).toBeInTheDocument();
        expect(btn).toBeInTheDocument();
        expect(input).toMatchSnapshot();
    });

    test('Async', async () => {
        render(<App/>);
        screen.debug();
        const dataElement = await screen.findByText(/data/i);
        expect(dataElement).toBeInTheDocument();
        expect(dataElement).toHaveStyle({color: 'red'});
        screen.debug();
    });

    test('Event click', async () => {
        render(<App/>);
        const toggleBtn = screen.getByTestId('toggle-btn');
        expect(screen.queryByTestId('toggle-elem')).toBeNull();
        fireEvent.click(toggleBtn);
        expect(screen.queryByTestId('toggle-elem')).toBeInTheDocument();
        fireEvent.click(toggleBtn);
        expect(screen.queryByTestId('toggle-elem')).toBeNull();
        fireEvent.click(toggleBtn);
        expect(screen.queryByTestId('toggle-elem')).toBeInTheDocument();
    });

    test('Input Fire Event', async () => {
        render(<App/>);
        const input = screen.getByTestId('input-elem');
        expect(screen.queryByTestId('value-elem')).toContainHTML('');
        fireEvent.input(input, {
            target: {value: '123123'}
        });
        expect(screen.queryByTestId('value-elem')).toContainHTML('123123');
    });

    test('Input User Event', async () => {
        render(<App/>);
        const input = screen.getByTestId('input-elem');
        expect(screen.queryByTestId('value-elem')).toContainHTML('');
        userEvent.type(input, '123123')
        expect(screen.queryByTestId('value-elem')).toContainHTML('123123');
    });
})

