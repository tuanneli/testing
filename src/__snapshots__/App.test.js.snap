// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App renders learn react link 1`] = `
<input
  placeholder="Input value..."
  type="text"
/>
`;

exports[`renders learn react link 1`] = `
<input
  placeholder="Input value..."
  type="text"
/>
`;
